src/binary-writer.cc

```diff
@@ -1121,6 +1286,31 @@ void BinaryWriter::WriteExpr(const Func* func, const Expr* expr) {
       break;
     }
   }
+  return std::nullopt;
+}
+
+void BinaryWriter::WriteExpr(const Func* func, const Expr* top_expr) {
+  std::list<StateVar> stack;
+
+  auto write_expr = [&stack, func, this](const Expr* expr) {
+    auto opt_state = WriteExprImpl(func, expr);
+    if (opt_state.has_value()) {
+      auto& state = *opt_state;
+      stack.push_back(state);
+    }
+  };
+  write_expr(top_expr);
+
+  while (!stack.empty()) {
+    State* state =
+        &std::visit([](auto& state) -> State& { return state; }, stack.back());
+    auto expr = state->advance(*this, func);
+    if (expr) {
+      write_expr(expr.value());
+    } else {
+      stack.pop_back();
+    }
+  }
```

Comment:
`(WriteInitExpr | WriteFunc) -> WriteExprList -> (WriteExpr | WriteExprList)*`
```suggestion
void BinaryWriter::WriteExprListImpl(const Func* func, const ExprList& exprs) {
  std::deque<StateVar> stack;
  WriteExprList(stack, func, exprs);
  while (not stack.empty()) {
    auto expr = std::visit([&](auto& state) { return state.advance(*this, func); }, stack.back());
    if (expr) {
      WriteExpr(stack, expr.value());
    } else {
      stack.pop_back();
    }
  }
```